src/storage.rs

```diff
@@ -115,32 +136,58 @@ impl<T: Clone + 'static> RingBufferStorage<T> {
         } else {
             self.written - reader_id.written
         };
+
+        let read_index = reader_id.read_index;
+        reader_id.read_index = self.write_index;
+        reader_id.written = self.written;
+
         if num_written > self.max_size {
-            let mut d = self.data
-                .get(self.write_index..self.max_size)
-                .unwrap()
-                .to_vec();
-            d.extend(self.data.get(0..self.write_index).unwrap().to_vec());
-            reader_id.read_index = self.write_index;
-            reader_id.written = self.written;
-            return Err(RBError::LostData(d, num_written - self.max_size));
+            Err(RBError::LostData(
+                StorageIterator {
+                    storage: &self,
+                    current: self.write_index,
+                    end: self.write_index,
+                    started: false,
+                },
+                num_written - self.max_size,
+            ))
+        } else {
+            Ok(StorageIterator {
+                storage: &self,
+                current: read_index,
+                end: self.write_index,
+                // handle corner case no data to read
+                started: num_written == 0,
+            })
         }
-        let read_data = if self.write_index >= reader_id.read_index {
-            self.data
-                .get(reader_id.read_index..self.write_index)
-                .unwrap()
-                .to_vec()
+    }
+}
+
+/// Iterator over a slice of data in `RingbufferStorage`.
+pub struct StorageIterator<'a, T: 'a> {
+    storage: &'a RingBufferStorage<T>,
+    current: usize,
+    end: usize,
+    // needed when we should read the whole buffer, because then current == end for the first value
```

Comment:
Mostly I see this corner case handled by leaving one index in buffer untouched. Hence `start == end` means there is nothing to read. And  `(start + 1) % capacity == end` means there is no place to push.
It removes necessity of boolean flags, simplify code and overall less error prone.

Reply:
Doesn't work, because the buffer knows nothing about how much readers have read.

Reply:
Should work. Just start read with `write_index + 1` if `num_written > self.data.len() - 1`